gateway/handle_http_auth.go

```diff
@@ -119,6 +119,44 @@ func (gw *Handle) sourceIDAuth(delegate http.HandlerFunc) http.HandlerFunc {
 	}
 }
 
+// authenticateDestinationIdForSource middleware to authenticate destinationId in the X-Rudder-Destination-Id header.
+// If the destinationId is invalid, the request is rejected.
+// destinationID authentication should be performed only after source is authenticated and source is present in context
+// Following validations are performed
+//  1. Destination should be present for source config
+//  2. Destination should be enabled for the source
+func (gw *Handle) authenticateDestinationIdForSource(delegate http.HandlerFunc) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		reqType := r.Context().Value(gwtypes.CtxParamCallType).(string)
+		arctx := r.Context().Value(gwtypes.CtxParamAuthRequestContext).(*gwtypes.AuthRequestContext)
+		var errorMessage string
+		defer func() {
+			gw.handleHttpError(w, r, errorMessage)
+			gw.handleFailureStats(errorMessage, reqType, arctx)
+		}()
+		destinationID := r.Header.Get("X-Rudder-Destination-Id")
+		if destinationID == "" {
+			if !gw.config.GetBool("Gateway.isDestinationIdMandatoryInRequestHeader", false) {
+				delegate.ServeHTTP(w, r)
+				return
+			}
+			errorMessage = response.NoDestinationIdInHeader
+			return
+		}
+		isValidDestination, isDestinationEnabled := gw.validateDestinationID(destinationID, arctx)
+		if !isValidDestination {
+			errorMessage = response.InvalidDestinationID
```

Comment:
Shouldn't this be DestinationNotFound instead of InvalidDestinationID?
```suggestion
			errorMessage = response.DestinationNotFound
```

Reply:
can destination not found can be considered as a case of Invalid destination, similar to what we are doing for Source ID not found?